### PR TITLE
Implemented an advanced types verification for the TypedStorage in the NativeContracts.

### PR DESCRIPTION
**Pull request by RomanZacharia**
Implemented an advanced types verification for the TypedStorage in the NativeContracts.

### CODE REVIEW

File: hydrachain/native_contracts.py [hunk 1]
```diff
@@ -714,9 +728,33 @@ def __getitem__(self, i):
...
         self.set(bytes(i), v)
+        self.updatelen(i, v)
+
+    def updatelen(self, i, v):
         if i >= len(self):
             self.set(b'__len__', i + 1, value_type='uint32')
+        elif i+1 == len(self) and v is 0:
+            self.set(b'__len__', 0, value_type='uint32')
+            # there is no good way to recalculate this, so iterate for now, but think of a better solution
+            for k in reversed(range(i)):
```
Comment (heikoheiko) [inline hydrachain/native_contracts.py]: this is potentially super slow. i'd trade speed for memory.

Comment (RomanZacharia) [inline hydrachain/native_contracts.py]: This change was reverted
https://github.com/RomanZacharia/hydrachain/blob/838029b53e1b29063296e1264c1002808b2b5fed/hydrachain/native_contracts.py#L744

**Comment by heikoheiko** [comment]
Can you elaborate a bit on what the problem was and how it was solved?

**Comment by RomanZacharia** [comment]
@heikoheiko 
1) The initial commit "Fixed length recalculation after elements removal from the List typed… " was addressing your request to fix the length recalculation for IterableDict and List, but later both were removed as you already implemented the recalculation for the ItedableDict on your own and the List's recalculation was removed due to the perfomance concerns;
2) The commit "Implemented an advanced types verification for the TypedStorage in th…". Since the TypedStorage-derived classes are strongly typed, their misusage was possible without the developer even noticing that since they did not issue any errors on the incorrect initialization attempts. The wrong data was just written to the wrong places in the storage. To address this I added the type checks on assignment and the following unit tests:
    a = nc.Dict(nc.List('uint16'))
    c = nc.List(nc.Dict('uint16'))
    k = nc.List(nc.List('address'))

```
a.setup(b'a', _get, _set)
c.setup(b'c', _get, _set)
k.setup(b'k', _get, _set)

# test invalid types

with pytest.raises(AttributeError):
    a.b == 81

with pytest.raises(ValueError):
    a['one']['two'] = 63432

with pytest.raises(TypeError):
    a['one'][2] = 'somestr'

with pytest.raises(NotImplementedError):
    k[1] = 2

with pytest.raises(NotImplementedError):
    c[1] = 2
```

3) The commit "Removed the tester.disable_logging() calls as obsolete" is pretty obvious, such method was removed from the tester.
4) The commit "Fixed an error in test_native_contracts.test_nac_tester()" addressed a Failing Travis build by adding the error handling for abi encoding.
5) The commit "Resolved the HydraChain#70 issue: "Reading Scalar from Struct in native contracts throws EncodingError" fixed calling the Struct getters/setters instead of the nested Scalar ones. So now the nested Scalar accesses the storage itself.

**Comment by konradkonrad** [comment]
py27 build succeeded -- merging

**Pull request merged**